GitLab Self-managed integration (#7856)
circleci/circleci-docs: jekyll/_cci2/gitlab-integration.adoc

You can choose a sample `.circleci/config.yml` to best fit your project from a variety of templates (for example, Node.js, Python, iOS apps). You can make changes to the file in CircleCI before saving. The config is committed on a new branch.

The first two options (**Fastest** and **Faster**) automatically trigger a pipeline once you create the project.
--

[.tab.signup.GitLab_Self-Managed]
--
You need:

> * Your GitLab self-managed instance URL, e.g. `https://example.circleci.com``.

Reviewer: ```suggestion
* Your GitLab self-managed instance URL, for example, `https://example.circleci.com``.
```

+
Your self-managed instance must already contain at least one project. The authorization attempt will be unsuccessful if your instance does not have any projects.

* A link:https://docs.gitlab.com/ee/user/profile/personal_access_tokens.html[personal access token]. This token must have the `api` scope.  

* Your instance's SSH key fingerprint. You can copy this from your instance's `known_hosts` file, which is located by default in the `~/.ssh/` directory.

To create a new project:

. Enter your self-managed instance URL. The field will automatically append `/api/v4`, so the URL that CircleCI uses to connect to your account looks something like `https://example.circleci.com/api/v4`.

. Enter your personal access token. Click **Connect**.

. Copy and paste the instance SSH key fingerprint into **Known hosts file**. Click **Connect**.

. Once your account is successfully authorized, select the repository for your CircleCI project, and enter a project name. 

. Finally, click **Create Project**. You will then be redirected to the CircleCI projects dashboard. 

> The express CircleCI configuration setup is not currently available for GitLab Self-managed projects. You will need to add a `.circleci/config.yml` file in your repository if it has not yet been set up with a CircleCI config. If the repository you selected already contains a `.circleci/config.yml`, you will need to save a commit in the repo to see your pipeline running in the dashboard.

Reviewer: ```suggestion
The express CircleCI configuration setup is not currently available for GitLab Self-managed projects. You will need to add a `.circleci/config.yml` file in your repository if it has not yet been set up. If the repository you selected already contains a `.circleci/config.yml`, you will need to save a commit in the repo to see your pipeline on the dashboard.
```

Reviewer: Just a terminology suggestion - pipelines are 'triggered" not "run", only jobs and workflows "run" ... this is buried in the style guide somewhere

# gitlab.com:22 SSH-2.0-GitLab-SSHD
gitlab.com ecdsa-sha2-nistp256 AAAAE2VjZHNhLXNoYTItbmlzdHAyNTYAAAAIbmlzdHAyNTYAAABBBFSMqzJeV9rUzU4kWitGjeR4PWSa29SPqJ1fVkhtj3Hw9xjLVXVYrU9QlYWrOLXBpQ6KWjbjTDTdDkoohFzgbEY=
# gitlab.com:22 SSH-2.0-GitLab-SSHD
gitlab.com ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIAfuCHKVTjquxvt6CM6tdG4SLp1Btn/nOeHHE5UOzRdf
# gitlab.com:22 SSH-2.0-GitLab-SSHD
# gitlab.com:22 SSH-2.0-GitLab-SSHD
```

To create a new project:

. Enter your self-managed instance URL. The field will automatically append `/api/v4`, so the URL that CircleCI uses to connect to your account looks something like `https://test-gitlab.circleci.com/api/v4`.

. Enter your personal access token. Click **Connect**.

. Copy and paste the SSH host key from your self-managed instance into **known_hosts**. Click **Connect**.  

. Once your account is successfully authorized, select the repository for your CircleCI project, and enter a project name. 

> . Finally, click **Create Project**. You will then be redirected to the CircleCI projects dashboard. 

Reviewer 2: **nit/non-blocking** We redirect them to the newly created project's pipelines page instead of the dashboard which has pipelines for all the projects they are following 🙂

* Your GitLab self-managed instance URL, e.g. `https://example.circleci.com``.
+
Your self-managed instance must already contain at least one project. The authorization attempt will be unsuccessful if your instance does not have any projects.  Note that the self-managed instance must be accessible via the public internet, **it cannot be behind a firewall**.

* A link:https://docs.gitlab.com/ee/user/profile/personal_access_tokens.html[personal access token]. This token must have the `api` scope.  

> * Your instance's SSH public host keys. You can retrieve this from your instance by running `ssh-keyscan <instance_url>`, for example, `ssh-keyscan test-gitlab.circleci.com`, and copying the command's output.

Reviewer 3: can we make this its own linkable section so that we can have a direct link from the in-app intructions?

Reviewer: It is! but there's something odd going on with the html conversion, I compared the JSON and the HTML and the id is ending up as `#user-content-known-hosts-input` somewhere along the way. We can use that anchor to link to for now, and I'll make a ticket to look into the addition of `user-content` as a bug in the platform I think

Reviewer: @rsylvian would you agree this is a platform thing/bug?

* A link:https://docs.gitlab.com/ee/user/profile/personal_access_tokens.html[personal access token]. This token must have the `api` scope.  

* Your instance's SSH key fingerprint. You can copy this from your instance's `known_hosts` file, which is located by default in the `~/.ssh/` directory.

>   ** To get this SSH key, use your self-managed instance URL and run `ssh-keyscan gitlab.com/yourInstanceURL`. The output should look something like:

Author: @sebastian-lerner The instance URL would look something like `example.circleci.com`, correct? Is it OK to change the URL example below as well?

Reviewer 3: i think itll look like gitlab.example.com because it's a gitlab self-managed instance?

[#trigger-pipeline]
== Trigger a pipeline in CircleCI

[.tab.pipeline.GitLab.com]
--
When you create a new project using the **Fastest** (use an existing `config.yml`) or **Faster** (commit a starter CI pipeline) options described in the section above, a pipeline is automatically triggered. You should see the pipeline running shortly after you are taken to the CircleCI dashboard.

If you use the **Fast** config setup, the pipeline is not triggered until you save the `.circleci/config.yml` by clicking the **Commit and Run** button in the web app.
--

[.tab.pipeline.GitLab_Self-Managed]
--
If you have not already done so, add a `.circleci` directory to the root of the repository, then add a `config.yml` file in that directory.

> When you commit this change in your repository, you should see the pipeline run for the first time on the CircleCI dashboard.

Reviewer: ```suggestion
When you commit this change in your repository, you should see the pipeline trigger for the first time on the CircleCI dashboard.
```

[#configuration]
=== Configuration

Currently, you can add or delete a configuration source for your project. If you followed the steps above to connect GitLab, a GitLab configuration source has been automatically added for you. 

> For GitLab Self-managed, you are able to select any instance that you have previously added as a configuration source. if you wish to set a different feature branch or repository from a self-managed instance as a new configuration source, you will first need to add a new connection via your xref:#organization-settings-integrations[**Organization Settings**]. In either case, you will also need to enter your personal access token again to authorize this connection.

Reviewer: ```suggestion
For GitLab Self-managed, you are able to select any instance that you have previously added as a configuration source. If you wish to set a different feature branch or repository from a self-managed instance as a new configuration source, you will first need to add a new connection via your xref:#organization-settings-integrations[**Organization Settings**]. In either case, you will also need to enter your personal access token again to authorize this connection.
```

+
If they do not currently have a CircleCI account, they will need to sign up. If they already have a CircleCI account, they are added to the organization, and if they are logged in, they will see the organization as an option in the organization switcher in the top left corner of the web app.

[#organization-settings-integrations]
=== Integrations (GitLab Self-managed only)

> For GitLab Self-managed organizations, you may connect additional self-managed instances to be integrated with your organization. Navigate to **Integrations** section within **Organization Settings** to add a new instance. You will need to enter the instance URL, your personal access token, and the SSH key fingerprint as described in the xref:#sign-up[Sign up] section earlier in this document.

Reviewer 2: We won't be asking for a personal access token on this page. 

Based on the current design
If users have already added the `known_hosts` then they will be able to add an instance URL. [Designs](https://www.figma.com/file/yWoX7C7KShtPaA5BgxXSLK/GL-self-managed-connection?node-id=658-11928&t=PdGGo884VLk8Xnpn-0)

<img width="659" alt="image" src="https://user-images.githubusercontent.com/22925321/229543643-8e985bfa-3346-428e-8518-1818951cf64a.png">

<img width="540" alt="image" src="https://user-images.githubusercontent.com/22925321/229541480-e44e61bc-6d0d-4385-a244-7fc1091fe8b3.png">.

If they have not added the `known_hosts` and have no connections they will see a `Set up Integration` button which will open a modal that will let them add the `known_hosts`. [Designs](https://www.figma.com/file/yWoX7C7KShtPaA5BgxXSLK/GL-self-managed-connection?node-id=652-10380&t=PdGGo884VLk8Xnpn-4)

<img width="683" alt="image" src="https://user-images.githubusercontent.com/22925321/229543071-a014e5c3-8d00-437e-8e79-8e5423ec8749.png">

<img width="312" alt="image" src="https://user-images.githubusercontent.com/22925321/229543120-33a7e3c6-5588-4b29-90b0-93e175c3685f.png">

> For GitLab Self-managed organizations, you may connect additional self-managed instances to be integrated with your organization. Navigate to **Integrations** section within **Organization Settings** to add a new instance. You will need to enter the instance URL, your personal access token, and the SSH key fingerprint as described in the xref:#sign-up[Sign up] section earlier in this document.

Reviewer: ```suggestion
For GitLab Self-managed organizations, you may connect additional self-managed instances to be integrated with your organization. Navigate to **Integrations** within **Organization Settings** to add a new instance. You will need to enter the instance URL, your personal access token, and the SSH key fingerprint as described in the xref:#sign-up[Sign up] section above.
```

> NOTE: The ability to edit or delete existing integrations is not currently enabled.

Reviewer: ```suggestion
NOTE: The ability to edit or delete existing integrations is not currently supported.
```

+
If they do not currently have a CircleCI account, they will need to sign up. If they already have a CircleCI account, they are added to the organization, and if they are logged in, they will see the organization as an option in the organization switcher in the top left corner of the web app.

> [#organization-settings-integrations]

Reviewer: A screenshot here would be great too

=== Integrations (GitLab Self-managed only)

For GitLab Self-managed organizations, you may connect additional self-managed instances to be integrated with your organization. 

. Navigate to **Integrations** within **Organization Settings** to add a new instance. 

> . You will need to enter the instance URL, and the SSH host keys if you have not already done so, as described in the xref:#sign-up[Sign up] section above. You have successfully set up the integration when you see a "Known hosts file verified" message.

Reviewer 2: **nit/non-blocking** We don't currently have this, but will be adding it soon. At the moment, we only have the option to add a new instance
